### PR TITLE
Add user model, new user input page, and user show page

### PR DESCRIPTION
**Pull request by gitlev**
Ropes Issue 12 Part 1

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [ ] I have replaced `X` above with the associated issue number.
- [ ] I have changed the issue's label to "pull request".
- [ ] I have assigned the designated reviewer to this PR.
- [ ] I have set the appropriate label on this PR.
- [ ] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: app/validators/email_validator.rb [hunk 1]
```diff
@@ -1,6 +0,0 @@
-class EmailValidator < ActiveModel::EachValidator
```
Comment (bboe) [inline app/validators/email_validator.rb]: Where did this come from? What is used for something else?

File: config/routes.rb [hunk 2]
```diff
@@ -7,4 +7,5 @@
       post 'send_email'
     end
   end
+  resources :users, only: [:new]
```
Comment (bboe) [inline config/routes.rb]: Does this change belong in this commit?

File: test/controllers/images_controller_test.rb [hunk 3]
```diff
@@ -45,7 +45,7 @@ class ImagesControllerTest < ActionController::TestCase
                                 share_form: { email_address: 'aperson',
                                               message: 'Look this pic' } },
                       xhr: true
-    assert JSON.parse(@response.body)['content'].include? 'Please enter a valid email address'
+    assert JSON.parse(@response.body)['content'].include? 'is invalid'
```
Comment (bboe) [inline test/controllers/images_controller_test.rb]: This seems like it belongs in its own commit along with changing to the gem for email validation.

File: test/models/user_test.rb [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                     password: 'foobar',
+                     password_confirmation: 'foobar')
+  end
+
+  test 'should be valid' do
+    assert_predicate @user, :valid?
+  end
+
+  test 'name should be present' do
+    @user.name = '     '
```
Comment (bboe) [inline test/models/user_test.rb]: The use of `setup` and an instance variable certainly works. However, consider defining a `create_user` private factory that returns a user object and uses keyword arguments whose defaults are valid.

In the way you can do something like  `user = create_user(name: '       ')`

File: test/models/user_test.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    assert_predicate @user, :invalid?
```
Comment (bboe) [inline test/models/user_test.rb]: For each failure case, please also assert the error messages that should be associated with the model. In this case you could start with `assert_equal {}, @user.errors.messages` (this might not be the perfect syntax, but it should be a start).

File: test/models/user_test.rb [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    valid_addresses.each do |valid_address|
+      @user.email = valid_address
+      assert @user.valid?, "#{valid_address.inspect} should be valid"
+    end
+  end
+
+  test 'email addresses should be unique' do
+    duplicate_user = @user.dup
+    duplicate_user.email = @user.email.upcase
+    @user.save
```
Comment (bboe) [inline test/models/user_test.rb]: What happens if `save` fails?

File: test/models/user_test.rb [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  end
+
+  test 'email validation should accept valid addresses' do
+    valid_addresses = %w(user@example.com USER@foo.COM A_US-ER@foo.bar.org
+                         first.last@foo.jp alice+bob@baz.cn)
+    valid_addresses.each do |valid_address|
+      @user.email = valid_address
+      assert @user.valid?, "#{valid_address.inspect} should be valid"
+    end
+  end
```
Comment (bboe) [inline test/models/user_test.rb]: While you don't need to exhaust all possible invalid emails, you should test at least one to ensure that your email validator is actually set up.

File: test/models/user_test.rb [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    duplicate_user = @user.dup
+    duplicate_user.email = @user.email.upcase
+    @user.save
+    assert_predicate duplicate_user, :invalid?
+  end
+
+  test 'email addresses should be saved as lower-case' do
+    mixed_case_email = 'Foo@ExaMPLe.CoM'
+    @user.email = mixed_case_email
+    @user.save
```
Comment (bboe) [inline test/models/user_test.rb]: What happens if save fails?

File: test/models/user_test.rb [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+  test 'password should be present (nonblank)' do
+    @user.password = @user.password_confirmation = ' ' * 6
+    assert_predicate @user, :invalid?
+  end
+
+  test 'password should have a minimum length' do
+    @user.password = @user.password_confirmation = 'a' * 5
+    assert @user, :invalid?
+  end
+end
```
Comment (bboe) [inline test/models/user_test.rb]: Things that I did not see in tests:

* Name of length 50
* Email of length 255
* Not email (already commented on)

File: app/views/users/show.html.erb [hunk 10]
```diff
@@ -2,6 +2,7 @@
   <aside class="col-md-4">
     <section class="user_info">
       <h1>
+        <%= gravatar_for @user %>
```
Comment (bboe) [inline app/views/users/show.html.erb]: This commit is a little confusing because github orders commits by their timestamps and not the parent-child-grandchild type order. If you reorganize commits it's a good idea to touch the timestamps of all the commits so that they show up in the "correct" order on github.

See: http://stackoverflow.com/a/31540373/176978

File: app/controllers/users_controller.rb [hunk 11]
```diff
@@ -0,0 +1,25 @@
+class UsersController < ApplicationController
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
+    if @user.save
+      flash[:success] = 'Welcome to the Sample App!'
+      redirect_to @user
```
Comment (bboe) [inline app/controllers/users_controller.rb]: Not that you can combine the previous line with redirect to:

    redirect_to @user, flash: {success: '...'}

File: app/controllers/users_controller.rb [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+    else
+      render 'new'
```
Comment (bboe) [inline app/controllers/users_controller.rb]: Is this a success?

File: app/views/users/new.html.erb [hunk 13]
```diff
@@ -0,0 +1,15 @@
+<h1>Sign up</h1>
```
Comment (bboe) [inline app/views/users/new.html.erb]: How does one get to this form?

image

File: app/views/users/show.html.erb [hunk 14]
```diff
@@ -0,0 +1,9 @@
+<div class="row">
```
Comment (bboe) [inline app/views/users/show.html.erb]: What's the purpose of the user show page?

File: test/controllers/users_controller_test.rb [hunk 15]
```diff
@@ -0,0 +1,26 @@
+require 'test_helper'
+
+class UsersControllerTest < ActionDispatch::IntegrationTest
+  test 'should get new' do
+    get new_user_url
+    assert_response :success
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: Please avoid using `:success` as it encompasses all 2XX level response statuses. What response status do you actually want to see?

Comment (bboe) [inline test/controllers/users_controller_test.rb]: Every one of your controller tests should have some check to ensure the response body has the important information. What is crucial to the new user view that you should use `assert_select` to check for?

File: test/controllers/users_controller_test.rb [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+  end
+
+  test 'invalid signup information' do
+    assert_no_difference 'User.count' do
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: It's great that you're checking for the side effect, or lack of, here, but what other important bits of response information are you not testing for?

File: test/controllers/users_controller_test.rb [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+      post users_path, params: { user: { name: '',
+                                         email: 'user@invalid',
+                                         password: 'foo',
+                                         password_confirmation: 'bar' } }
+    end
+  end
+
+  test 'valid signup information' do
+    assert_difference 'User.count', 1 do
+      post users_path, params: { user: { name: 'Example User',
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: It's great that you're checking for the side effect here, but what other important bits of response information are you not testing for?

File: app/views/users/show.html.erb (hunk 14 above)
Comment (bboe) [inline app/views/users/show.html.erb]: Any response to the above question?

File: test/controllers/users_controller_test.rb [hunk 18]
```diff
@@ -0,0 +1,37 @@
+require 'test_helper'
+
+class UsersControllerTest < ActionDispatch::IntegrationTest
+  test 'should get new' do
+    get new_user_url
+    assert_select "form[action='/users'][method='post']"
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: Great assertions. For simplicity, I think this line is sufficient to ensure you are loading the right form.

File: test/controllers/users_controller_test.rb [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+
+  test 'invalid signup information' do
+    assert_no_difference 'User.count' do
+      post users_path, params: { user: { name: '',
+                                         email: 'user@invalid',
+                                         password: 'foo',
+                                         password_confirmation: 'bar' } }
+    end
+    assert_response :unprocessable_entity
+    assert_select '.user_name.has-danger' do
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: Great assertion for the side effect.

File: test/controllers/users_controller_test.rb [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+  test 'valid signup information' do
+    assert_difference 'User.count', 1 do
+      post users_path, params: { user: { name: 'Example User',
+                                         email: 'user@example.com',
+                                         password: 'password',
+                                         password_confirmation: 'password' } }
+    end
+    assert_redirected_to user_path(User.last)
+  end
+end
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: Since you've added a show controller, there are at least two test cases needed.

File: test/models/user_test.rb [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+    user1 = build(:user)
+    user2 = build(:user)
+    user1.email = user2.email.upcase
+    user1.save
+    assert_predicate user2, :invalid?
+  end
+
+  test 'email addresses should be saved as lower-case' do
+    mixed_case_email = 'Foo@ExaMPLe.CoM'
+    user = create(:user, email: mixed_case_email)
```
Comment (gitlev) [inline test/models/user_test.rb]: It outputs an error and returns nil as far as I can tell

File: app/views/users/show.html.erb (hunk 14 above)
Comment (bboe) [inline app/views/users/show.html.erb]: > are you saying I should redirect to index?

I think it's good to try to keep things as simple as possible unless you have a specific reason to do more. Even if you want to do more, consider adding it in later so that you can keep the task at-hand as simple as possible.

File: test/controllers/users_controller_test.rb (hunk 15 above)
Comment (bboe) [inline test/controllers/users_controller_test.rb]: As I think I mentioned in the subsequent reply, the main thing that is important is that the form renders with the right action.

File: test/models/user_test.rb [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+      user.email = invalid_address
+      assert user.invalid?, "#{invalid_address.inspect} should be invalid"
+    end
+  end
+
+  test 'email addresses should be unique' do
+    user1 = build(:user)
+    user2 = build(:user)
+    user1.email = user2.email.upcase
+    user1.save
```
Comment (bboe) [inline test/models/user_test.rb]: Yes a system failure can occur. In this case it would silently return false if there were a system failure, or if somehow the user you are trying to create is invalid.

In rails there is both `save` and `save!`, the latter will raise an exception if the save fails. Consider using that instead wherever you are not explicitly checking the return value. The same applies for `create` and `create!`.

File: test/models/user_test.rb (hunk 21 above)
Comment (bboe) [inline test/models/user_test.rb]: Is there a way to get it to raise an exception if it doesn't result in the test explicitly failing here. If not perhaps doing `build(...).save!` is a better option.

File: app/controllers/users_controller.rb [hunk 23]
```diff
@@ -0,0 +1,32 @@
...
+
+  def user_params
+    params.require(:user).permit(:name, :email, :password, :password_confirmation)
+  end
+
+  def find_user_or_redirect
+    if (@user = User.find_by(id: params[:id]))
+      @user
+    else
+      flash[:danger] = 'The user profile you are trying to access could not be found.'
```
Comment (bboe) [inline app/controllers/users_controller.rb]: Awesome! Good work. Sometimes its okay to simply let it fail as RecordNotFound will raise and result in translation to a 404. But if you want to provide the user a slightly better experience this is a solid approach.

File: test/controllers/users_controller_test.rb [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+    assert_response :ok
+    assert_select '.user_info' do
+      assert_select 'h1', 'John Doe'
+      assert_select 'h1' do
+        assert_select "img[alt='John Doe'][class='gravatar']"
+      end
+    end
+  end
+
+  test 'show invalid user' do
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: Great test.

File: test/controllers/users_controller_test.rb [hunk 25]
```diff
@@ -0,0 +1,54 @@
...
+    # get :show, params: { id: -1 }
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: Do you think these comments add much to the code?

File: test/models/user_test.rb (hunk 21 above)
Comment (gitlev) [inline test/models/user_test.rb]: is this just to prevent unpredictability in our tests?

Comment (bboe) [inline test/models/user_test.rb]: In general if you call a method whose return value indicates success or failure, then you should be explicitly checking that value and not assuming it works. So, no, it's not just for unpredictability, it's to help promote writing good code.
